### DISCUSSION TITLE
Wimboot specifying wim index not working in version 2.7.4

### DISCUSSION THREAD

**Discussion by rushyn** [ipxe/wimboot#40] [Q&A] [answered]
Has any one else run in to an issue after updating to last wimboot(2.7.4) where index= (number) no longer works, and instead it boots the first index of the wim file? Updated from wimboot 2.5.2 going back to 2.5.2 resumes normal behavior.

**Comment by mcb30** [comment]
Could you try using git bisection to find the precise commit that caused the problem?

You can find a guide for bisecting at https://ipxe.org/howto/bisect.  It's written for iPXE, but the idea is the same for wimboot: you'll want to do something like:
* `git clone https://github.com/ipxe/wimboot.git`
* `make -C src`
* `git bisect start`
* `git bisect good v2.5.2`
* `git bisect bad v2.7.4`

and then repeatedly

* `make -C src`
* test the resulting wimboot binary
* `git bisect good` if it works, or `git bisect bad` if it fails

until `git` reports that it has identified the culprit commit.

**Reply by rushyn** [reply]
I have run bisect result below, Let me know if you want the log.

In addition I have remove the 4 lines of code added in commit ac4190adb505d081c7bc41ffac392af0e7627d26 from last commit 2240d7f0396eb574ac789b968f76bf9e54bbf3d8.

Now it boots the correct index of the wim. Note that I do not have the skill to figure out if removing this lines brock something else.

Last thing I had said that it was always booting the first index of the wim, that is not correct, it was always booting the last index of the wim.


------------------------------------------
commit ac4190adb505d081c7bc41ffac392af0e7627d26
Author: Michael Brown <mbrown@fensystems.co.uk>
Date:   Thu Feb 11 02:27:14 2021 +0000

    [patch] Avoid patching wimboot binary itself into image
    
    When booting on a UEFI system, the "wimboot" binary will itself be
    visible within the underlying EFI_SIMPLE_FILE_SYSTEM_PROTOCOL.
    
    Avoid injecting this file into the \Windows\System32 directory of the
    .wim image.
    
    Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

 src/wimpatch.c | 4 ++++
 1 file changed, 4 insertions(+)
------------------------------------------

**Reply by mcb30** [reply]
@rushyn Thank you for running the bisection.  Could you please try the following:

* Boot using the unmodified commit https://github.com/ipxe/wimboot/commit/2240d7f0396eb574ac789b968f76bf9e54bbf3d8 with an `index=...` command-line parameter to specify a non-default index and an additional `pause` command-line parameter
* Boot using your modified commit https://github.com/ipxe/wimboot/commit/2240d7f0396eb574ac789b968f76bf9e54bbf3d8  with the four lines from commit https://github.com/ipxe/wimboot/commit/ac4190adb505d081c7bc41ffac392af0e7627d26 removed, with the same `index=... pause` command-line parameters

That should get you a screenshot showing the modifications performed to the WIM image.  Could you please post both screenshots here?

**Reply by rushyn** [reply]
@mcb30 I had no issues taking a picture of the unmodified wimboot output, however once I removed the lines from src/wimpatch.c the out put was too long to fit on one screen, I have edited together two video frames to capture the full output.

This is the wimboot line "kernel ftp://192.168.1.254/wimboot pause index=50"

Here is the out put for unmodified wimboot.

![image](https://user-images.githubusercontent.com/65286397/213958422-cc72098e-9bc0-4628-ae00-0569fb46e913.png)

This is the output for the same commit with the 4 lines removed.

![image](https://user-images.githubusercontent.com/65286397/213960149-2706fe78-9224-4d68-b9b9-35b934bf418f.png)

**Reply by mcb30** [reply]
Thank you.  Your screenshot has coincidentally revealed an unrelated bug: https://github.com/ipxe/wimboot/issues/41 :slightly_smiling_face: 

For your issue: I do see the `...patching WIM header at [0x0,0xd0)` in both screenshots, and this is the only part that should be relevant in terms of patching the boot index.

Could you please try the unmodified commit https://github.com/ipxe/wimboot/commit/2240d7f0396eb574ac789b968f76bf9e54bbf3d8 but built using `make DEBUG=3` to enable additional debugging?  The additional debug messages that I am interested in are the ones produced by:
https://github.com/ipxe/wimboot/blob/2240d7f0396eb574ac789b968f76bf9e54bbf3d8/src/wimpatch.c#L246-L273

These messages will not appear until `bootmgfw.efi` starts reading from the patched WIM image, which happens only after the `Press any key to continue booting...` prompt (if enabled).  They may be challenging to capture via video: if so, you should be able to safely add an extra call to `pause()` at the end of `wim_patch_header()`:

```patch
--- a/src/wimpatch.c
+++ b/src/wimpatch.c
@@ -269,6 +269,9 @@ static int wim_patch_header ( struct wim_patch *patch,
        }
        memcpy ( data, ( ( ( void * ) &patch->header ) + offset ), len );
 
+       // temporary hack
+       pause();
+
        return 0;
 }
 
```

**Reply by rushyn** [reply]
Happy I could help find a bug another bug, 

I did not have any luck getting a good video of the of the debug output, After adding the pause() to wimpatch.c the compiler errored out.
![image](https://user-images.githubusercontent.com/65286397/214178429-b5b5f42a-2eaa-4b48-9dc9-129dac96e47a.png)

Removed the -Werror flag from make file, and it complied with this warranting. 
![image](https://user-images.githubusercontent.com/65286397/214178601-1861f0a5-ac0c-432a-be00-e69002f22498.png)

Is this what you are after?

![image](https://user-images.githubusercontent.com/65286397/214179015-c9d00412-7d7d-448b-b00f-9a57aebac6c8.png)

**Reply by mcb30** [reply]
> Is this what you are after?
> 
> ![image](https://user-images.githubusercontent.com/65286397/214179015-c9d00412-7d7d-448b-b00f-9a57aebac6c8.png)

Yes, thank you.  I believe that gives me enough information to figure out what is happening.  I'm going to convert this to an issue since it is a genuine bug.

**Comment by mcb30** [comment, marked as answer]
Moving to issue #43 (and locking this discussion: please continue conversation in the issue)
